### PR TITLE
*: change schema from []*Column to struct

### PR DESCRIPTION
**Pull request by winoros**
i do this because i will add unique key information into schema.
PTAL @coocood @hanfei1991 @shenli @zimulala @lamxTyler

### CODE REVIEW

File: executor/adapter.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func (a *statement) Exec(ctx context.Context) (ast.RecordSet, error) {
 	}
 
 	// Fields or Schema are only used for statements that return result set.
-	if len(e.Schema()) == 0 {
```
Comment (hanfei1991) [inline executor/adapter.go]: add a Length() function to schema

File: expression/column.go [hunk 2]
```diff
@@ -151,9 +151,18 @@ func (col *Column) HashCode() []byte {
...
 func (col *Column) ResolveIndices(schema Schema) {
-	col.Index = schema.GetIndex(col)
+	col.Index = schema.GetColumnIndex(col)
 	// If col's index equals to -1, it means a internal logic error happens.
 	if col.Index == -1 {
 		log.Errorf("Can't find column %s in schema %s", col, schema)
 	}
 }
+
+// Column2Exprs will transfer column array to expression array.
```
Comment (hanfei1991) [inline expression/column.go]: slice not array

File: expression/expression.go [hunk 3]
```diff
@@ -261,22 +261,22 @@ func ResultFieldsToSchema(fields []*ast.ResultField) Schema {
...
 			ColName:  col.Name,
 			TblName:  tbl.Name,
 			RetType:  &col.FieldType,
 			Position: i,
 		}
-		schema = append(schema, newCol)
+		cols = append(cols, newCol)
 	}
-	return schema
+	return Schema{Columns: cols}
```
Comment (hanfei1991) [inline expression/expression.go]: Add a NewSchemaByCols function

File: expression/schema.go [hunk 4]
```diff
@@ -61,40 +64,37 @@ func (s Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
...
 
-//Schema2Exprs converts []*Column to []Expression.
-func Schema2Exprs(schema Schema) []Expression {
-	result := make([]Expression, 0, len(schema))
-	for _, col := range schema {
-		result = append(result, col.Clone())
-	}
-	return result
+// MergeSchema will merge two schema into one schema.
+func MergeSchema(lSchema, rSchema Schema) Schema {
```
Comment (hanfei1991) [inline expression/schema.go]: Merge Schema should change the argument itself. It's better to create a new schema

File: plan/planbuilder.go [hunk 5]
```diff
@@ -602,76 +602,76 @@ func (b *planBuilder) buildExplain(explain *ast.ExplainStmt) Plan {
...
-	schema = append(schema, buildColumn(tblName, "Timing", mysql.TypeVarchar, 128))
-	schema = append(schema, buildColumn(tblName, "Created", mysql.TypeDatetime, 19))
-	schema = append(schema, buildColumn(tblName, "sql_mode", mysql.TypeBlob, 8192))
-	schema = append(schema, buildColumn(tblName, "Definer", mysql.TypeVarchar, 128))
-	schema = append(schema, buildColumn(tblName, "character_set_client", mysql.TypeVarchar, 32))
-	schema = append(schema, buildColumn(tblName, "collation_connection", mysql.TypeVarchar, 32))
-	schema = append(schema, buildColumn(tblName, "Database Collation", mysql.TypeVarchar, 32))
-	return schema
+	colSchema := make([]*expression.Column, 0, 11)
+	colSchema = append(colSchema, buildColumn(tblName, "Trigger", mysql.TypeVarchar, 128))
```
Comment (hanfei1991) [inline plan/planbuilder.go]: Implement Append Function for Schema

File: expression/schema.go (hunk 4 above)
Comment (coocood) [inline expression/schema.go]: Then we don't need to call `Clone` on original schemas.

**Comment by shenli** [comment]
What is the benefit of adding unique key into schema?

**Comment by winoros** [comment]
Thus we could do some optimization based on the unique key information.

**Comment by shenli** [comment]
LGTM
@winoros Please solve the conflict.

File: expression/schema.go [hunk 6]
```diff
@@ -61,40 +63,51 @@ func (s Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
...
 
-//Schema2Exprs converts []*Column to []Expression.
-func Schema2Exprs(schema Schema) []Expression {
-	result := make([]Expression, 0, len(schema))
-	for _, col := range schema {
-		result = append(result, col.Clone())
-	}
-	return result
+// GetColumnsLen returns the number of columns in schema.
+func (s Schema) GetColumnsLen() int {
```
Comment (hanfei1991) [inline expression/schema.go]: name `Len` is enough.

File: expression/schema.go [hunk 7]
```diff
@@ -61,40 +63,51 @@ func (s Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
...
+	return len(s.Columns)
+}
+
+// AppendColumn append new column to the columns stored in schema.
+func (s *Schema) AppendColumn(col *Column) {
```
Comment (hanfei1991) [inline expression/schema.go]: name `Append` is enough.

**Pull request merged**